## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [5]:
# What three item categories had the most entries?
df.value_counts('category').head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries,
# which subcategory had the most entries?
categories_df = df.loc[df['category']=='consumables']
categories_df.value_counts('subcategory').head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
df.value_counts('client_id').head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
client_df = df.value_counts('client_id')
client_list = (client_df.iloc[:5]).index.tolist()
client_list

[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the
# client with the most entries order order?
client_max = client_list[0]
quantity_df = df.loc[df['client_id']==client_max]
quantity_df['qty'].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['Subtotal'] = df['unit_price'] * df['qty']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   Subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

def shipping_price(x):
    if x > 50:
        return 7*x
    return 10*x

df['Total_weight'] = df['qty'] * df['unit_weight']
df['Shipping_Price'] = df['Total_weight'].apply(shipping_price)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   Subtotal  Total_weight  \
0     762.71         7.50  105            1  115164.00        787.50   
1      15.09         1.49   21            0     523.95         31.29   
2       7.86         1.68   39            6     527.28         65.52   
3      24.85         1.23   29            3    1056.18         35.67   
4     108.17        46.43   20            1    3902.00        928.60   

   Shipping_Price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [12]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = (df['Subtotal'] + df['Shipping_Price']) * 1.0925
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   Subtotal  Total_weight  Shipping_Price  \
0         7.50  105            1  115164.00        787.50         5512.50   
1         1.49   21            0     523.95         31.29          312.90   
2         1.68   39            6     527.28         65.52          458.64   
3         1.23   29            3    1056.18         35.67          356.70   
4        46.43   20            1    3902.00        928.60         6500.20   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 22 columns]

In [13]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['total_cost'] = (df['unit_cost']*df['qty']) + df['Shipping_Price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   Subtotal  Total_weight  Shipping_Price    total_price  \
0            1  115164.00        787.50         5512.50  131839.076250   
1            0     523.95         31.29          312.90     914.258625   
2            6     527.28         65.52          458.64    1077.117600   
3            3    1056.18         35.67          356.70    1543.571400   
4            1    3902.00        928.60         6500.20   11364.403500   

   total_cost  
0    85597.05  
1      629.79  
2      765.18  
3     1077.35  
4     8663.60  

[5 rows x 23 columns]

In [14]:
# Create a column for the profit of
# each line using line cost and line price
df['Profit']= df['total_price']-df['total_cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

    Subtotal  Total_weight  Shipping_Price    total_price  total_cost  \
0  115164.00        787.50         5512.50  131839.076250    85597.05   
1     523.95         31.29          312.90     914.258625      629.79   
2     527.28         65.52          458.64    1077.117600      765.18   
3    1056.18         35.67          356.70    1543.571400     1077.35   
4    3902.00        928.60         6500.20   11364.403500     8663.60   

         Profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
order_ID_2742071 = df.loc[df['order_id']==2742071]
total_2742071 = order_ID_2742071['total_price'].sum()
print(f'Order ID 2742071 had a total price of ${total_2742071:,.2f}')

order_ID_2173913 = df.loc[df['order_id']==2173913]
total_2173913 = order_ID_2173913['total_price'].sum()
print(f'Order ID 2173913 had a total price of ${total_2173913:,.2f}')

order_ID_6128929 = df.loc[df['order_id']==6128929]
total_6128929 = order_ID_6128929['total_price'].sum()
print(f'Order ID 6128929 had a total price of ${total_6128929:,.2f}')

Order ID 2742071 had a total price of $152,811.90
Order ID 2173913 had a total price of $162,388.72
Order ID 6128929 had a total price of $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
sorted_df = df.sort_values(by='qty', ascending=False)
top_client_ids = sorted_df['client_id'][:5]
for client_id in top_client_ids:
    client_data = sorted_df.loc[sorted_df['client_id'] == client_id]
    total_qty = client_data['qty'].sum()
    total_spending = client_data['total_price'].sum()
    print(f"Client ID: {client_id}")
    print(f"Total Quantity: {total_qty:,.2f}")
    print(f"Total Spending: ${total_spending:,.2f}")
    print("-" * 30)

Client ID: 78965
Total Quantity: 4,033,302.00
Total Spending: $655,923,670.06
------------------------------
Client ID: 86721
Total Quantity: 1,010,426.00
Total Spending: $181,266,297.07
------------------------------
Client ID: 30143
Total Quantity: 846,069.00
Total Spending: $33,328,849.83
------------------------------
Client ID: 85931
Total Quantity: 644,423.00
Total Spending: $237,417,001.42
------------------------------
Client ID: 77354
Total Quantity: 506,382.00
Total Spending: $71,942,651.78
------------------------------


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_data = []
for client_id in top_client_ids:
    client_data = sorted_df.loc[sorted_df['client_id'] == client_id]
    total_units = client_data['qty'].sum()
    total_shipping = client_data['Shipping_Price'].sum()
    total_revenue = client_data['total_price'].sum()
    total_profit = client_data['Profit'].sum()
    summary_data.append([client_id, total_units, total_shipping, total_revenue, total_profit])
summary_df = pd.DataFrame(summary_data, columns=['Client ID', 'Total Units', 'Total Shipping', 'Total Revenue', 'Total Profit'])
summary_df = summary_df.sort_values(by='Total Profit', ascending=False)
summary_df

Client ID  Total Units  Total Shipping  Total Revenue  Total Profit
0      78965      4033302      5340756.88   6.559237e+08  2.473528e+08
3      85931       644423     10122549.19   2.374170e+08  1.092941e+08
1      86721      1010426     64335727.72   1.812663e+08  4.297470e+07
4      77354       506382     23814685.86   7.194265e+07  1.243226e+07
2      30143       846069      8981905.24   3.332885e+07  1.125039e+07

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
summary_df.columns = ['Client', 'Units Purchased', 'Shipping Cost', 'Revenue', 'Profit']
million_formatter = lambda x: f"${x / 1000000:,.2f}M"
summary_df['Shipping Cost'] = summary_df['Shipping Cost'].apply(million_formatter)
summary_df['Revenue'] = summary_df['Revenue'].apply(million_formatter)
summary_df['Profit'] = summary_df['Profit'].apply(million_formatter)
summary_df

Client  Units Purchased Shipping Cost   Revenue    Profit
0   78965          4033302        $5.34M  $655.92M  $247.35M
3   85931           644423       $10.12M  $237.42M  $109.29M
1   86721          1010426       $64.34M  $181.27M   $42.97M
4   77354           506382       $23.81M   $71.94M   $12.43M
2   30143           846069        $8.98M   $33.33M   $11.25M

In [19]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df

Client  Units Purchased Shipping Cost   Revenue    Profit
0   78965          4033302        $5.34M  $655.92M  $247.35M
3   85931           644423       $10.12M  $237.42M  $109.29M
1   86721          1010426       $64.34M  $181.27M   $42.97M
4   77354           506382       $23.81M   $71.94M   $12.43M
2   30143           846069        $8.98M   $33.33M   $11.25M